File: py/customerioapp_sdk/features.py
# CustomerioApp SDK feature factory

from customerioapp_sdk.feature.base_feature import CustomerioAppBaseFeature
from customerioapp_sdk.feature.debug_feature import CustomerioAppDebugFeature
from customerioapp_sdk.feature.idempotency_feature import CustomerioAppIdempotencyFeature
from customerioapp_sdk.feature.metrics_feature import CustomerioAppMetricsFeature
from customerioapp_sdk.feature.paging_feature import CustomerioAppPagingFeature
from customerioapp_sdk.feature.ratelimit_feature import CustomerioAppRatelimitFeature
from customerioapp_sdk.feature.retry_feature import CustomerioAppRetryFeature
from customerioapp_sdk.feature.test_feature import CustomerioAppTestFeature
from customerioapp_sdk.feature.timeout_feature import CustomerioAppTimeoutFeature


_FEATURES = {
    "base": lambda: CustomerioAppBaseFeature(),
    "debug": lambda: CustomerioAppDebugFeature(),
    "idempotency": lambda: CustomerioAppIdempotencyFeature(),
    "metrics": lambda: CustomerioAppMetricsFeature(),
    "paging": lambda: CustomerioAppPagingFeature(),
    "ratelimit": lambda: CustomerioAppRatelimitFeature(),
    "retry": lambda: CustomerioAppRetryFeature(),
    "test": lambda: CustomerioAppTestFeature(),
    "timeout": lambda: CustomerioAppTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES

File: php/features.php
<?php
declare(strict_types=1);

// CustomerioApp SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class CustomerioAppFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new CustomerioAppBaseFeature();
            case "debug":
                return new CustomerioAppDebugFeature();
            case "idempotency":
                return new CustomerioAppIdempotencyFeature();
            case "metrics":
                return new CustomerioAppMetricsFeature();
            case "paging":
                return new CustomerioAppPagingFeature();
            case "ratelimit":
                return new CustomerioAppRatelimitFeature();
            case "retry":
                return new CustomerioAppRetryFeature();
            case "test":
                return new CustomerioAppTestFeature();
            case "timeout":
                return new CustomerioAppTimeoutFeature();
            default:
                return new CustomerioAppBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}
